src/tree/RefreshManager.ts

```diff
@@ -0,0 +1,255 @@
+import { AllTreePrefixes, TreePrefix } from './TreePrefix';
+
+const pollingIntervalMs = 60 * 1000; // One minute
+const eventListenerTries = 3; // The event listener will try at most 3 times to connect for events
+
+type RefreshTarget = AzExtTreeItem | TreePrefix;
+type RefreshReason = 'interval' | 'event' | 'config' | 'manual' | 'contextChange';
+
+const ContainerEventActions: EventAction[] = ['create', 'destroy', 'die', 'kill', 'pause', 'rename', 'restart', 'start', 'stop', 'unpause', 'update'];
+const ImageEventActions: EventAction[] = ['delete', 'import', 'load', 'pull', 'save', 'tag', 'untag'];
+const NetworkEventActions: EventAction[] = ['create', 'destroy', 'remove'];
+const VolumeEventActions: EventAction[] = ['create', 'destroy'];
+
+export class RefreshManager extends vscode.Disposable {
+    private readonly disposables: vscode.Disposable[] = [];
+    private readonly cts: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+
+    public constructor() {
+        super(() => vscode.Disposable.from(...this.disposables).dispose());
+
+        // VSCode does *not* cancel by default on disposal of a CancellationTokenSource, so we need to manually cancel
+        this.disposables.unshift(new vscode.Disposable(() => this.cts.cancel()));
+
+        this.setupRefreshOnInterval();
+        this.setupRefreshOnRuntimeEvent();
+        this.setupRefreshOnConfigurationChange();
+        this.setupRefreshOnCommand();
+        this.setupRefreshOnDockerConfigurationChange();
+        this.setupRefreshOnContextChange();
+    }
+
+    private setupRefreshOnInterval(): void {
+        const timer = setInterval(async () => {
+            for (const view of AllTreePrefixes) {
+                // Skip the registries view, which does not need to be refreshed on an interval
+                if (view === 'registries') {
+                    continue;
+                }
+                await this.refresh(view, 'interval');
+            }
+        }, pollingIntervalMs);
+
+        this.disposables.push(new vscode.Disposable(
+            () => clearInterval(timer)
+        ));
+    }
+
+    private setupRefreshOnRuntimeEvent(): void {
+        void callWithTelemetryAndErrorHandling('vscode-docker.tree.eventRefresh', async (context: IActionContext) => {
+            context.errorHandling.suppressDisplay = true;
+            context.telemetry.suppressIfSuccessful = true;
+
+            const eventTypesToWatch: EventType[] = ['container', 'image', 'network', 'volume'];
+            const eventActionsToWatch: EventAction[] = Array.from(new Set<EventAction>([...ContainerEventActions, ...ImageEventActions, ...NetworkEventActions, ...VolumeEventActions]));
+
+            // Try at most `eventListenerTries` times to (re)connect to the event stream
+            for (let i = 0; i < eventListenerTries; i++) {
+                try {
+                    const eventGenerator = await ext.streamWithDefaultShell(client =>

```

Comment:
Can the events come faster than we can process them? Should we "debounce" the refresh invocations?

Reply:
They do come pretty fast in certain circumstances and we probably should debounce. The net effect if we don't is that we spawn a couple of CLI calls at once, which can return in an arbitrary order, but since we just update the data to whatever is returned it isn't harmful (aside from the wasted CPU time).

I'm still deciding what the best debouncing approach is. There's basically two ways I can think of:

1. Ignore events of the same type that come in within X milliseconds of the last one
2. Collapse rapidly-fired events of the same type into one refresh that occurs X milliseconds after the last one (more or less how we do diagnostics in the compose language service--diagnostics get sent ~1 second after the user stops typing)

The disadvantage of the first strategy is that we could miss relevant changes to the data that occur after the first call; the disadvantage of the second is the implementation is a little more complex and it adds latency before real changes are reflected in the UI. I think the second approach is still better though.

Reply:
Agreed. Thanks for good explanation